File: evd_authoring_app/src/model/evdScript/program/machineOperations/machinePrimitive.js
import { Primitive } from '../primitive';

export class MachinePrimitive extends Primitive {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'machine-primitive';
    }

    static fullTypeString() {
        return Primitive.fullTypeString() + MachinePrimitive.typeString();
    }

    constructor(machineUuid=null, type='', name='', uuid=null, parent=null, appendType=true) {
        super(
            (appendType) ? 'machine-primitive.'+type : type,
            name,
            uuid,
            parent,
            appendType
        );

        this._machineUuid = null;

        this.machineUuid = machineUuid;
    }

    toDict() {
        const msg = {
            ...super.toDict(),
            machine_uuid: this.machineUuid
        };
        return msg;
    }

    static fromDict(dct) {
        return new MachinePrimitive(
            dct.machine_uuid,
            dct.type,
            dct.name,
            dct.uuid,
            null,
            false
        );
    }

    /*
    * Data accessor/modifier methods
    */

    get machineUuid() {
        return this._machineUuid;
    }

    set machineUuid(value) {
        if (this._machineUuid !== value) {
            this._machineUuid = value;
            this.updatedAttribute('machine_uuid','set');
        }
    }

    set(dct) {

        if ('machine_uuid' in dct) {
            this.machineUuid = dct.machine_uuid;
        }

        super.set(dct);
    }

    /*
    * Update methods
    */

    deepUpdate() {
        super.deepUpdate();

        this.updatedAttribute('machine_uuid','update');
    }

    shallowUpdate() {
        super.shallowUpdate();

        this.updatedAttribute('machine_uuid','update');
    }
}

File: archive/evd_ros_backend/evd_ros_core/src/evd_script/program_nodes/machine_primitive.py
'''
Machine Primitive extends primitive to simplify the later definition of specific machine primitives.

A machine primitive is a way to generalize machine behavior in EvD. The actual behavior needs to be
supplied externally and is hooked into the EvD program runner.
'''

from .primitive import Primitive
from ..data_nodes.machine import Machine


class MachinePrimitive(Primitive):

    '''
    Data structure methods
    '''

    @classmethod
    def display_name(cls):
        return 'Machine Primitive'

    @classmethod
    def type_string(cls, trailing_delim=True):
        return 'machine-primitive' + ('.' if trailing_delim else '')

    @classmethod
    def full_type_string(cls):
        return Primitive.full_type_string() + cls.type_string()

    @classmethod
    def template(cls):
        template = Primitive.template()
        template['parameters'].append({
            'type': Machine.full_type_string(),
            'key': 'machine_uuid',
            'is_uuid': True,
            'is_list': False
        })
        return template

    def __init__(self, machine_uuid=None, parameters=None, type='', name='', uuid=None, parent=None,
                 append_type=True, editable=True, deleteable=True, description=''):

        if parameters == None:
            parameters = {
                'machine_uuid': machine_uuid
            }

        super(MachinePrimitive,self).__init__(
            type=MachinePrimitive.type_string() + type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type,
            editable=editable,
            deleteable=deleteable,
            description=description,
            parameters=parameters)

    '''
    Data accessor/modifier methods
    '''

    @property
    def machine_uuid(self):
        return self.parameters['machine_uuid']

    @machine_uuid.setter
    def machine_uuid(self, value):
        if self.parameters['machine_uuid'] != value:
            self.parameters['machine_uuid'] = value
            self.updated_attribute('parameters.machine_uuid','set')

    def set(self, dct):
        if 'machine_uuid' in dct.keys():
            self.machine_uuid = dct['machine_uuid']

        super(MachinePrimitive,self).set(dct)
